apps/meteor/server/publications/messages.ts

```diff
@@ -7,65 +7,230 @@ import type { FindOptions } from 'mongodb';
+	options: FindOptions<IMessage>;
+} {
+	const options: FindOptions<IMessage> = {
+		...(!next && { sort: { ts: -1 } }),
+		...(next || previous ? { limit: count + 1 } : {}),
+	};
+
+	if (next) {
+		return { query: { $gt: extractTimestampFromCursor(next) }, options };
+	}
+
+	if (previous) {
+		return { query: { $lt: extractTimestampFromCursor(previous) }, options };
+	}
+
+	return { query: { $gt: new Date(0) }, options };
+}
+
+function mountCursorFromMessage(message: IMessage & { _deletedAt?: Date }, type: 'UPDATED' | 'DELETED'): string {
+	if (type === 'UPDATED') {
+		return `${message._updatedAt.getTime()}`;
+	}
+
+	if (type === 'DELETED' && message._deletedAt) {
+		return `${message._deletedAt.getTime()}`;
+	}
+
+	throw new Meteor.Error('error-cursor-not-found', 'Cursor not found', { method: 'messages/get' });
+}
+
+function mountNextCursor(messages: IMessage[], type: CursorPaginationType, next?: string, previous?: string): string | null {
+	if (next) {
+		messages.reverse();
+	}
+
+	// if messages length is greater than 0, it means that we might have more messages
+	// and we shall return a cursor pointing to the most recent message from the current batch
+
+	// since we are already sorted by descending order on database query
+	// it's not performant to enumerate the existence of a next cursor
+	// so we can just return the timestamp from the most recent message
+	// and the client will move forward in the messages
+	if (messages.length > 0) {
+		return mountCursorFromMessage(messages[0], type);
+	}
+
+	if (messages.length === 0 && previous) {
+		return `${parseInt(previous, 10) - 1}`;
+	}
+
+	return null;
+}
+
+function mountPreviousCursor(messages: IMessage[], type: CursorPaginationType, count?: number, next?: string): string | null {
+	// if count is null, we don't have to return a cursor
+	// because we are going to return all the messages
+	if (count === null) {
+		return null;
+	}
+
+	// if messages length is 0 and next is provided, it means we reached the end of the messages
+	// and we shall return a cursor that inverts the current query
+	if (messages.length === 0 && next) {
+		return `${parseInt(next, 10) + 1}`;
+	}
+
+	// if messages length is equal to count and next is provided, it means we reached the end of the messages
+	// and we shall return a cursor that inverts the current query
+	if (messages.length === count && next) {
+		return `${parseInt(next, 10) - 1}`;
+	}
+
+	// if messages length is less or equal to count, it means we are at the end of the messages
+	// and we shall not return a cursor
+	if (count && messages.length <= count) {
+		return null;
+	}
+
+	// if messages length is greater than count, it means we are not at the end of the messages
+	// and we shall return a cursor
+	if (count && messages.length > count) {
+		return mountCursorFromMessage(messages[messages.length - 2], type);
+	}
+
+	return null;
+}
+
+function popMessagesFromArray(messages: IMessage[], count?: number) {
```

Comment:
I think we could pop the message directly there, since it's not being reused, we should not create this function.